### ISSUE TITLE
IMPORTANT: Migrating to modern javascript/typescript

### ISSUE THREAD

**Issue by asturur** [fabricjs/fabric.js#7596]
Hello everyone!

FabricJS is nearly 12 years old.

The library has still its place around the internet, is good for many things, is never perfect, but i know that for who approach rich canvas based interactive apps, this is a nice piece of software to leverage.

While our main fault are still lack of great docs and updated and clarifying examples, now also the javascript code looks like a lot obsolete.

There are things we could leverage like Promises, default value for arguments, fat arrow for loops and readibility, native classes, proper getters and setters.

We would like to take the occasion to revisit the shape of the codebase to something more modern.
This would include proably a bundler, proper import statements instead of a single file that gets built like lego blocks, maybe tree shaking for who import just what he wants to import.

This would make the custom built unnecessary, using other libraries easier including touch interactions.

So a move to es201X is coming anyway.

The question is if we want to add type descriptors or if we want to make a TS migration right away.

The benefits of consuming TS are clear to me, the benefit of writing typescript a little less, i'm not a great fan and also i m not great a TS.
Is very easy to use TS in the wrong way and produce subpar results.

We are open to suggestions and understand what is the best way to make everyone happy.
external type definition in form of d.ts files are a good middle way if there is a way to enforce correctness at build time.
A new lint profile is required.
A new build tool too.
Legacy JS support is not going away.
A bundler will produce the correct code for older browsers.
But probably IE11 is out of the game anyway. I would love to have 2 builds, a legacy one and a modern one ( the classic last 2 versions of each browser )

So which are your reason to list TS on the pro/cons of this change?
I understand you may not like TS. that is fine. I don't like it either.
But if is terse, with types out of the way, and a transpiled build is available, what is your main point against or for it?

**Comment by melchiar** [comment]
@jimmywarting Typescript support is one of our most commonly requested features. Feel free to elaborate if you have specific objections to adding support for it.

**Comment by jimmywarting** [comment]
esm is also a desirable request, working with typescript and esm is a total nightmare. beside you can have a type safety system with vanila js + jsdoc and writing the code in such a manner that it don't require annotation... like using # instead of private, use classes instead of prototype, avoid using promise constructor so the IDE can really figure out what something return and so on.

it's also possible to generate d.ts files from js also

i created https://jimmywarting.github.io/you-might-not-need-typescript/ to this end...

**Comment by andrewzolotukhin** [comment]
I think it's not necessary to migrate to Typescript,  but it would be great to have up to date d.ts typings for Fabric.

**Comment by asturur** [comment]
i was supposed to fill up the issue text. I ll do soon. We are looking for suggestion, we are all ears and we don't have strong opinions yet.

We would like to move away from legacy js anyway.

**Comment by ShaMan123** [comment]
I am for TS (gave a read to what you wrote @jimmywarting, looks interesting).
IMO the disadvantage of .d.ts files is maintenance. Because it is not part of the code it can quickly become outdated and irrelevant.
Personally speaking I find TS most helpful in replacing the need for docs in my workflow. 
My IDE suggests whatever TS has to offer and my learning curve grows exponentially because of it.
Bottom line, that's what won me over, besides the trivial (great 😅) advantages.
Regarding the type nightmare - I understand what you mean but I would expect the pain to reduce after the first adoption phase. I can counter with my pains from js. From the back of my head, not knowing function signatures and having to guess, remember them or log arguments to move forward. Is JSDOC typedef visible outside of it's file/module? If not it is a killer.
From the maintainer point of view I think it cuts down Q&A as well.
So I'm for.
I do think that a simple configuration file (tsconfig.json) in a project can silence ts for a dev that doesn't want it.
And it is the standard as far as I can tell (saw many large repros migrate to ts, including google open source).
I think that when considering this move we should try to foresee what is right for 1-2 years ahead, not only for today.
In other words, we need to think what is best for fabric in the long term (and hopefully the short as well)

**Comment by jimmywarting** [comment]
> IMO the disadvantage of .d.ts files is maintenance. Because it is not part of the code it can quickly become outdated and irrelevant.

this can be auto generated durning release...

> I think that when considering this move we should try to foresee what is right for 1-2 years ahead, not only for today.

what would you think could happen when [optional static typing](https://github.com/sirisian/ecmascript-types) comes into play? typescript will become obsolete.

> saw many large repros migrate to ts

just b/c big companies did it dosen't mean it's better than one or the other, it just comes to a preferences and the developers background (like coming from c, rust, go or any other typed language)

For me it feels like typescript are trying to fit a square into a circle, javascript is a dynamic language, and it's nothing wrong with that, it can be a positive thing also...

there are ways to write the code in such a way that auto completion and annotation isn't necessary from either jsdoc or typescript also. 
The public/private keyword and explicit return type for example are not needed, they are unnecessary, even for typescript. 
the only time you really need to explicit say what something returns is if you do something like:
```js
function sleepFoo() {
  return new Promise(rs => {
    setTimeout(() => {
      rs('bar')
    }, 100)
  })
}
```
which don't have to be annotated at all if you did:
```js
async function sleepFoo() {
  await sleep(100)
  return 'bar'
}
```

beside typescript don't do any type protection after you compile it to js.
ppl can still make mistake afterwards.
typescript only provide type hints/error upfront durning developing.

**Comment by Gerwin-prog** [comment]
@asturur will you also move away from supporting Legacy JS entirely & only support TypeScript?

**Comment by kirill-konshin** [comment]
Based on my experience, most of the projects can be migrated to TS with minimal effort (obviously, the types will not be very strict, some compromises will take place), but it's a proper foundation to improve these types in future.

TS can work even w/o types as plain JS, types can be introduced function by function, file by file.

Additionally, w/o TS project will have to support custom .d.ts files, which is a nightmare, it's much easier to convert code to TS and generate .d.ts automatically.

**Comment by keelii** [comment]
I am for TS too, because:

1. old style javascript sometimes is hard to understand(even though you are expert on JS for years). prototype extends, contextual-this.., there is so much lines are hard to read in fabric.js
```js
fabric.Object.prototype._setLineDash.call(this, ctx, this.selectionDashArray)
```
2. when i implement my custom object.class, the `createClass` method sometimes misleading people(ES 6 class looks more explicity), it hard to know which method should be implemented or overrided(the underscore method or not or both?)
```js
var MyClass = fabric.util.createClass(fabric.Object, {
    bar: [], // dont't do this
    initialize: function () {},
})
var c1 = new MyClass()
c1.bar.push(1)
var c2 = new MyClass()
console.log(c2.bar) // => [1]

var MyClass = fabric.util.createClass(fabric.Object, {
    initialize: function () {
        this.bar = [] // do this instead
    },
})
var c1 = new MyClass()
c1.bar.push(1)
var c2 = new MyClass()
console.log(c2.bar) // => []
```
3. there are some global variables in fabric namespace, which is hard to find(you have to read the source code to find it), with TS, just give me a interface, problem solved.
```js
fabric.disableStyleCopyPaste
fabric.copiedText
```
4. when you built a large canvas-based project, the code style always OOP right? TS bring the full OOP support to JS, when i coding with TS, I feel confident, enjoyable. with JS i always confused, it too dynamic. in fabric.js the actionHandler in `fabric.Control` class is a absolute beautiful abstraction for object transform action, but when i involved in this part of code, it's hard to find where i am.

**Comment by asturur** [comment]
I updated the description.
Probably the best option is to maybe convert a non working example of a class, to see how does it looks like in modern JS and TS.
Maybe RECT that is a quick one.

**Comment by jafferhaider** [comment]
I'm for vanilla JS, because we finally have decent traction on modern features being implemented in ES. Looking at the long-term, I feel it would be better for FabricJS to stay with a maturing vanilla JS because many new features on the roadmap will make TS irrelevant anyway. 

If we were having this discussion a half a decade ago, I would have been all for TypeScript.

**Comment by kirill-konshin** [comment]
Colleagues, main point about Typescript is it’s strong and flexible type system. No vanilla js (even with jsdoc) gives true type checking during compilation time. In all my JS projects migrated to TS, type checks found a number of potential issues that were overlooked.

Second reason is DTS (definition) files. Since we’re talking about library development, TS is a de facto standard. Definitions must be produced in any case, if we target large audience, which includes other TS projects.

Definitions on top of JS tend to have errors, mismatches and discrepancies, and require a lot of manual maintenance. Of course one can generate dts from jsdoc, but this will not do type checking… and in case JSDoc has issues those issues will silently leak into DTS, which means more maintenance efforts.

Third, TS is mature tool, with huge community, it’s safe and proven. Vanilla JS is catching up, it’s still years behind TS in terms of features. And even when it will catch up, vanilla won’t have types anyway, which brings us back to previous points.

**Comment by jimmywarting** [comment]
I don't consider any compile to js a "standard" language... It don't run anywhere without being compiled to js in the first place

It is built on top of standard js

**Comment by jimmywarting** [comment]
I like buildless setup a lot
It loads so insanely fast 🚀 
Having esm would be cool too

Only time when i actually use any build tool is if I need to down level anything for older environments 

Butt all browser basically have auto update now and I'm it's pretty dead...

**Comment by asturur** [comment]
which are the functions of TS that are better than JS, types apart? i thought where identical.
I wonder if we can get best of bot, we write in ts, and we can build into plain js and bundled legacy js

**Comment by jimmywarting** [comment]
> I wonder if we can get best of both, we write in TS, and we can build into plain JS and bundled legacy JS

- When you write in TS and build to JS then there isn't any plain vanilla javascript anywhere anymore... 
- When you write in TS you get the best of TypeScript and the worst of javascript, annoying sourcemaps, extra compile time, bloated code that looks nowhere near the actual input. most often you get one single bundled file that are not so friendly code splitted, One more more language to learn, soft private methods that should not be used (if you use public/private instead of #), broken path resolving (when it comes to using extension - for esm) stuck with one extra chain of development pipeline for getting the newest hottest features b/c you need to wait for TypeScript to catch up with new standard javascript specification. 
- TS can transform your working code to broken stuff that no longer work (did u see my ESM example?)

The only way you are going to get both of both world is if you write in JS and turn on checkJS to utilize all TypeScript features, you may even use typescript to bundle to legacy js form plain js
**jsdoc can do type checking**

VScode has a nice interfear type from usage that helps u fill in the jsdoc when necessary, you can write interfaces (typedef) and import them with jsdoc as well...

**Comment by asturur** [comment]
yes my initial idea was that strong jsdoc could generate d.ts files.
But i never found a proper tool to do it.

Isn't TS with ripped of types just plain JS?
There must be an option for the compiler, `don't touch js` no?

**Comment by asturur** [comment]
ok let's start to check `checkJS`

**Comment by jimmywarting** [comment]
> Isn't TS with ripped types just plain JS?

not when you have target set to like some very old environment, it's always going to do some form of transformation to your code. 1000 of developers think they write es syntax with their imports but the reality is that it most often is compiled to commonjs

> There must be an option for the compiler, don't touch js no?

[alloJS](https://www.typescriptlang.org/tsconfig/allowJs.html)?

**Comment by kirill-konshin** [comment]
@jimmywarting Your point "annoying sourcemaps, extra compile time, bloated code that looks nowhere near the actual input" contradicts with your other point "target set to like some very old environment, it's always going to do some form of transformation to your code" — you can't publish raw ESM code.

"all browser basically have auto update now and IE it's pretty dead" — Random old browsers still exist in corporate sector, where auto update may be turned off... Also don't forget government sector. You still have to compile CJS and maybe even ES5. Obviously we should not target IE6 or IE7, but only targeting latest Chrome is another extreme.

Fabric is a very popular lib, most popular based on stars, it cannot throw away compatibility, so there WILL be a build/compile script, one or the other.

![image](https://user-images.githubusercontent.com/1530410/149364396-8ece7b0a-cde0-4284-8230-faf0fd5fd5b6.png)

Competition is tough...

VSCode does have a good support for JSDoc, but I've never seen a reliable tool to check JSDoc types on CI, whereas TS compiler will do it during the build as normal part of CI process.

@asturur wrote:
> which are the functions of TS that are better than JS, types apart? i thought where identical.
> I wonder if we can get best of bot, we write in ts, and we can build into plain js and bundled legacy js
> Isn't TS with ripped types just plain JS?"

This is true, for the most part JS now is TS without types, except a few features like decorators (which Fabric may use as strong-typed mixin replacement). When TS is used you get types for free, DTS for external consumption.

**Comment by jimmywarting** [comment]
> you can't publish raw ESM code.

I could if i wanted to. I can code and develop in raw ESM, making sure it run just fine in latest chrome and maybe firefox, without having to compile anything.
And when i want to make a release _then_ i can run some test that builds some UMD, test it in several browsers and releases both ESM (maybe as is - for those who wish to simply import it) and a UMD version for the rest.

it's possible to use webpack, rollup or tsc or whatever.

**Comment by kirill-konshin** [comment]
@jimmywarting 
> I could if i wanted to. I can code and develop in raw ESM, making sure it run just fine in latest chrome and maybe firefox, without having to compile anything. And when i want to make a release _then_ i can run some test that builds some UMD

Yeah, you can do that, but you trade compilation time during dev at expense of testing and verification time. And again, you won't have to support DTS files separately... It's always a tradeoff. So unless you actually ship raw ESM I don't see practical benefits. I advocate for "Eating your own dog food" to be sure I am seeing what others will see, as close as it can be. And as early as I can, ideally during development, not verification.

Compilation time in watch mode of a library the size of fabric is few seconds. It's not a big deal. But then during development you can be sure the code you see is exactly what will be shipped. 

To me, when you develop a LIBRARY — TS is a no brainer choice. When you develop an APP — your approach makes more sense.

Also there are rust (SWC) or Go (esbuild) based compilers that are even faster than regular TSC/Babel/Rollup/etc., and looks like it's the general direction [where industry is moving](https://nextjs.org/blog/next-11-1#adopting-rust-based-swc), Next.js used to be a JS project, and they did full transition to TS, it's just one example... overall the trend is towards TS in lib development.

Overall the choice should be based on if team will be able to work faster with TS or with JS + custom DTS. From my experience in multiple teams and products, TS helps to speed up overall delivery if all factors are considered (development, dev experience, CI, QA, support).

**Comment by ShaMan123** [comment]
I think it is clear that fabric should not and cannot deal with the build process. It it unwise, to say the least, and a damn waste of effort.
The effort should be targeted towards fabric and making it awesome. Let everyone do what they do best.
And of course fabric must use a liable mechanism, no alpha/beta products. Too much headache in something that no one should have to manage except the occasional ci config. 
The patience to except alpha/beta and the occasional break/bug is **ONLY** intended for fabric features.
And It **MUST** make this repo easier and readable! Because it isn't!
I am an expert in fabric internals not because I want too. But because I had no choice but to dig in the source code and break my head against it. Why should anyone want to work with a repo that demands so much just to get some lame customization working? It's too hard, messy, horrible DX.
The point is taking fabric forward, not drowning it with more noise. No one wants that.

**Comment by asturur** [comment]
Yes but indeed there would be 2 targets eventually. One to get modern JS out of TS, another to have the standard browser ready lib

**Comment by asturur** [comment]
> I think it is clear that fabric should not and cannot deal with the build process. It is unwise, to say the least, and a damn waste of effort.

Well we don't get away without publishing a ready to consume module. Is what npm user expect.
You don't build stuff that is in your node_modules folder usually. You expect it to be ready to use.
If you want to use the modules right away you usually import them from the src directory.
There is not just a single way to do things of course, but i m sure we have to build.

**Comment by CaptEmulation** [comment]
Types would be amazing whether provided by declaration files or an actual Typescript port. Actual Typescript port and a build step is best long term because you can trust the types more.

**Comment by ShaMan123** [comment]
> > I think it is clear that fabric should not and cannot deal with the build process. It is unwise, to say the least, and a damn waste of effort.
> 
> Well we don't get away without publishing a ready to consume module. Is what npm user expect. You don't build stuff that is in your node_modules folder usually. You expect it to be ready to use. If you want to use the modules right away you usually import them from the src directory. There is not just a single way to do things of course, but i m sure we have to build.

Sure, we have to build. But with minimum effort. Let the build tool do all the work.

**Comment by andrewzolotukhin** [comment]
I think building is something we should avoid for Fabric. But we can definitely move towards ESM. I would not use TS, instead of it I'd create a d.ts files and add typings through JSDoc if needed.

**Comment by folknor** [comment]
For what it's worth, I want to voice my thoughts:
- I love the activity in fabricjs and I hope it continues to improve, and I think having the guts to make breaking changes like this is awesome and should be done more often
- I hope you can make a release before doing this, mostly so we can get the `jsdom` bump from v15 to v19. v15 chains to a lot of outdated dependencies
- If I was doing this, I would look long and hard at how bootstrap does it; https://github.com/twbs/bootstrap/tree/main/js. Note a few things;
1. All components are in separate files, with some shared utility classes
2. The original source is included in the npm distribution, so people can use their own build tools on it when including it in their project (and do their own tree shaking, minifying, etc)
3. They generate 3 versions using a simple `rollup` config (but other build utils can also easily be made to do this): https://github.com/twbs/bootstrap/blob/main/package.json#L40-L42
4. The resulting compiled JS is committed to git when they do a release, and is also part of the npm distribution: https://github.com/twbs/bootstrap/tree/main/js/dist
5. They do not provide their own TS definitions, they are maintained separately here https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/bootstrap

**Comment by kirill-konshin** [comment]
> They do not provide their own TS definitions, they are maintained separately here

I worked a lot with this project. Separate declaration is one of the reasons why bugs occur because of discrepancies. It is a very bad approach. Types should be shipped along with library.

> The resulting compiled JS is committed to git when they do a release

This is also considered as bad practice, build artifacts should not be part of version-controlled repo because it leads to horrific merge conflicts.

**Comment by kirill-konshin** [comment]
Here's an example that illustrates why Fabric has to be written in TS:

http://fabricjs.com/docs/fabric.Object.html#forEachControl
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/fabric/fabric-impl.d.ts — does not even mention this method `forEachControl`

And I doubt JSDOC to TS types generator will be a good solution: it is awkward and does not give all benefits that TS can provide.

**Comment by ShaMan123** [comment]
@kirill-konshin I have a huge patch file of types for fabric on top of the npm types module.
I agree it is bad.

**Comment by asturur** [comment]
i posted on defintely typed if someone wanted to help build official d.ts file to keep here, but i got no one to volunteer.
i m all for types comeing from here, the only thing at stake is between full TS and JS + .d.ts

**Comment by asturur** [comment]
@folknor release 5.0 is around the corner. I'm looking for the small last fixes to pull in.

**Comment by ShaMan123** [comment]
see the ts proto -
very simple, nothing fancy, 
minimal effort (promises including with a bit of a more advanced type related to the function):
https://github.com/ShaMan123/fabric.js/blob/ts-proto/src/shapes/Rect.ts
run `npx tsc` or `npm run tsc` after you've update dev deps. It will output a readable js file file next to the ts one.
Give it a go.

**Comment by ShaMan123** [comment]
https://www.typescriptlang.org/#adopt-gradually

**Comment by Devil7DK** [comment]
I'm also for TS. Most of the things I wanted to say to @jimmywarting's arguments @kirill-konshin already said it.
And I'll gladly help if you decide to migrate to TS.

**Comment by kingschnulli** [comment]
My main concern currently is that there is no tree shaking or real modules. For the TS/ES6/Plain or whatever point, we have babel so there is no reason to not just write everything the way @asturur and the other maintainers want. But still, will give +1 for TS because this will help everybody in the long way.

**Comment by asturur** [comment]
With those changes we would have proper modules, and custom build would go away.

**Comment by kingschnulli** [comment]
Whatever you will decide, let the contributors know how we can help in getting things done - I think there are quite a bunch of people willing to help for this transition.

**Comment by asturur** [comment]
So for example, help me figuring out, what is the best transpiler for library?
i work on apps and there webpack + babel is a standard, but i don't think is for libraries.
Can TS do transpilation and bundling all together?

**Comment by kingschnulli** [comment]
I'm using webpack to build a library which bundles current fabricjs and our custom vuejs code - so this is doable. But I think there is a lot going on with esbundle, swc, parcel etc. currently which might be a better solution. Maybe someone with more experience can chime in here and make a good proposal what bundler solution is the best for fabric.

**Comment by kirill-konshin** [comment]
You can use TSC compiler to build the library.

If you want to also publish the final build like this one https://cdn.jsdelivr.net/gh/fabricjs/fabric.js@vv6/dist/fabric.js you can do an extra step and compile it using Webpack with SWC loader (which is the fastest way to compile TS without types, since you don't need types in Webpack bundle).

Here's how I did it long time ago: https://github.com/ringcentral/ringcentral-js/blob/master/sdk/package.json

3 builds: ES5 for oldies, ES6 with tree-shaking for modern tools, UMD (webpack) for direct consumption.

**Comment by andrejslogins** [comment]
I'm also using fabric in the project I'm at currently, and personally, would love to see typescript used. Especially since now, I'm investigating some internal issue which feels like a race condition causing some of the variables being undefined when being accessed

**Comment by yassilah** [comment]
I personally also would love to see TS support. I have made a few (small) contributions to this repo and I found it quite hard to navigate the codebase without reliable parameters types, method return types and proper signatures. And it would really be my pleasure to help migrate to TS if there's anything you need 🙂

**Comment by Lazauya** [comment]
I'm definitely down to help work on the migration too; I just found a pretty big discrepancy between the DefinitelyTyped types and the actual lib, and so I think that migrating to TS would be pretty awesome.

**Comment by Lazauya** [comment]
So the question is, how do we want to approach this? Should we first use a tool to generate initial types and slowly fill stuff in from there?

Also, what should we do about the "Klass" stuff? Do we want to gut that and replace it with real modern classes?

**Comment by ShaMan123** [comment]
> So the question is, how do we want to approach this? Should we first use a tool to generate initial types and slowly fill stuff in from there?
> 

Any experience with reliable tools?

> Also, what should we do about the "Klass" stuff? Do we want to gut that and replace it with real modern classes?

**Absolutely!**

**Comment by Lazauya** [comment]
> Any experience with reliable tools?

Unfortunately no. I'm going to do some research to see what's available. Ideally a tool to add simple types like `string`, `bool`, etc. Manual typing could be done for the fabric objects. I'm planning on creating a branch with the basic types added from whatever tool I end up using and then going from there.

**Comment by yassilah** [comment]
Hi! I don't really understand the use of a tool to generate types tbh? I think the best way to go is migrate files from .js to .ts going from least dependency to most dependency (e.g. go from utils > mixins > shapes > collections > canvas or something like that). For each file, there may be a need for special interfaces to avoid repeating args, which we could declare in a `/types` root folder to allow access from other files. As for tools, we could think of using `ts-mixer`, which makes it quite easy to reuse mixins across classes. We could also include a small mixin utility to avoid any external dependency.

**Comment by ShaMan123** [comment]
Because it a big effort and many of us want to help we need to think and discuss how to break down the work so that everyone can take part in a good and useful way. We will need to make some kind of table assigning what to whom and to keep track of progress, maybe even use some teams tool.
We will definitely need to maintain a commons PR that will be updated frequently and everyone will have to update their branch all the time to stay in sync. And add some tools and how tos to make life easier, plus PR guideline and reviewing guidelines and tracker issues.
We must address all open PRs as well so that they won't die out in light of this major change.
I think we need a bit more time for preparations, to be honest.
This change is a chance to establish a plugin concept. Classes will be rewritten so we should take advantage of it.
I believe renaming will be done as well, as part of deciding what's private/public. Too many useful methods start with `_`.
Anyways, if you are working on something state it clearly ASAP so we can discuss it and to avoid unnecessary work. And perhaps target, as suggested, the independent modules first. POCs should precede major changes (that's for the guidelines 😉).
It might be a good idea for us all to target utils first, complete it and move on. This will require fabric to be in a transient mode which is possible I guess but a build tool will need to be setup in advance to support this.
Exciting times!

**Comment by ShaMan123** [comment]
BTW if any of you want to help in preparation there are some simple PRs that can be reviewed. It's a good way to understand how to develop fabric and use the dev tools. We can help.

**Comment by kirill-konshin** [comment]
FYI https://devblogs.microsoft.com/typescript/a-proposal-for-type-syntax-in-javascript/

Btw, you can migrate to TS by allowing plain JS and add compilation step to build. Once this is done you can start to transition types from JSDOC to TS file by file at any pace.

**Comment by Lazauya** [comment]
> Hi! I don't really understand the use of a tool to generate types tbh? I think the best way to go is migrate files from .js to .ts going from least dependency to most dependency (e.g. go from utils > mixins > shapes > collections > canvas or something like that). For each file, there may be a need for special interfaces to avoid repeating args, which we could declare in a `/types` root folder to allow access from other files. As for tools, we could think of using `ts-mixer`, which makes it quite easy to reuse mixins across classes. We could also include a small mixin utility to avoid any external dependency.

I've been doing research on existing tools, and it doesn't seem like there's anything that would fit this project's needs, so I'm thinking that this will end up being the best approach.

> BTW if any of you want to help in preparations there are some simple PRs that can be reviewed. It's a good way to understand how to use the dev tools. We can help.

I can start with this for sure. Any specific PR's?

**Comment by ShaMan123** [comment]
> I can start with this for sure. Any specific PR's?

#7714 #7713 #7689 #7805 
#7719
#7775 #7776 
 #7806 #7802

**Comment by Lazauya** [comment]
> > I can start with this for sure. Any specific PR's?
> 
> #7714 #7713 #7689 #7805 #7719 #7775 #7776 #7806 #7802

For newcomers like myself, what is the review process? Is it codified anywhere? Are these just simple code reviews?

**Comment by ShaMan123** [comment]
1. code review
2. clone PR
3. `npm run all` (runs tests and exports files to the websire directory that should reside nest to `fabric.js` directory)
4. check visual diffs by removing the failing files and running `npm run test:visual` 
5. start fabricjs.com and test the changes live

I am working on better dev tools as well so this process will become easier.

**Comment by pollow** [comment]
Not a hardcore JS expert but has been working on JS on and off for many years. I think with ~EMCAScript 2020, the latest JS, TS has been playing less important roles. For many times you can see, TS compiled results are identical to vanilla JS. And backward compatibility can be done by babel polypill. But since this is a new version with new standards, do we really need IE8- support?

As a 3P library, I would vote for less dependencies on tools & languages, but using vanilla JS to be closer to standard.

But type system is definitely a bit plus for any larger project, especially with more collaborations. FlowJS might be a good approaches, which is slightly more closer to the new JS optional type annotation proposal. And the type can be stripped off easily, be added progressively., and generate .d.ts by other tools in babel when release.

**Comment by ShaMan123** [comment]
I guess this should be discussed here...

As part of the migration I want to get rid of these:
- [ ] extend/clone in favor of lodash
- [ ] left/top in favor of x/y - this will be very easy once everything is written in TS. VSCode doesn't it in a heartbeat.
- [ ] shared object controls. There's no advantage in using it, only disadvantages.
**Look** at the loading times. You are welcome to compare against fabricjs.com No real diff.
![ezgif com-gif-maker](https://user-images.githubusercontent.com/34343793/164429071-7490849c-aea0-4e4e-a6ab-04cd1017c9ba.gif)

**Comment by asturur** [comment]
we can probalby stop using extended/clone with standard es6 tools.
I don't want to have lodash as a runtime dependency. We should not need to clone stuff at all that often.
Most of the options are clonable with `{ ...options }`.
We clone to defend against mutation, all the other situations should be avoided.

left/top vs x/y: x/y makes more sense, yes. Seems a big headache tho.

All the disavantages of shared controls is the textbox edgecase confusion, but you can prove me wrong.

**Comment by svjansen** [comment]
I've to say we're using fabric.js in a very large project with ES6+ and esm of course.
We had to rewrite a few little things and even got some extending classes here and there but dont have any problems to migrate everything we need.

However I can't see any reason why you should use TS in a rewrite. We're working with JetBrain tools and JSDoc and never had any kind of type problems. If we find something we need to "param guessing" we write a little wrapper class that got the exact same function names and just build JSDoc around.

**Comment by xorb** [comment]
Definitely go with typescript. Why not ? Type system is a great plus for any larger apps plus we can create a build that everyone can use it (js users, ts users,  etc).

**Comment by asturur** [comment]
oh the current index.js to build is transitory of course.
It is because we have this mixed situation in which we need to move the files to TS and i didn't want to have a PR that was renaming and changing all files and the build system.

When we are done, the import * from and export * from should do the job ok.
Then we have to figure out which is the best setup for custom builds.
the npm package should have the dist and the rest, so you can do something like
`import { Canvas, Text } from 'fabric'`
And maybe not having to import all fabric inside the app if you are doing just some text rendering.

**Comment by svjansen** [comment]
@asturur maybe you should keep in mind/should know, that Microsoft got a proposal on TC39 currently in Stage 1.
In short: they want to add static optional typing to JS in TypeScript format that should be ignored in runtime.

[MS DevBlog](https://devblogs.microsoft.com/typescript/a-proposal-for-type-syntax-in-javascript/)
[tc39 GitHub](https://github.com/tc39/proposal-type-annotations)
[tc39 official page](https://tc39.es/proposal-type-annotations/)

**Comment by asturur** [comment]
the proposal is actually typescript when i stop thinking.
The types never go to runtime, because no one wants to download or parse them at runtime, no one event wants the added code in the parsers to ignore them.
That is why JSDOCS is so good, it is typescript in comments, with descriptions and link to examples and other things.

Why they didn't write a type checker for jsdocs?

i m not a runtime expert, we all want performances for js, does the extra pass of parsing + stripping cost me something?
Does it make sense if then any minifier will strip them away for good?
I get they want to avoid sytnax conflicts with future ecmascript features so that the TS ecosystem doesn't get stuck with incompatibilities with es20XX.

**Comment by svjansen** [comment]
The only thing they say about JSdoc is [this](https://github.com/tc39/proposal-type-annotations#limits-of-jsdoc-type-annotations).
What they talk about over and over is that ppl trying to get rid of unnecessary build steps.

I can say that this would make things easier if you got multiple objects with same name where you would need to use namespace/modules syntax on JSDoc.
A Problem i have in JSDoc are types not instantiated in a js file itself. Lets say i got factory for an fabric.Image in an class named ImageHandler, but the ImageHandler itself never instantiate a fabric.Image. In this case it feels like luck if JSDoc and the IDE get the correct Object. In the proposal you would import fabric.Image just for typing to avoid such a Problem.

**Comment by ShaMan123** [comment]
#8020 exposes a mechanism that is in charge of transforming `src` files into es6 syntax (and convert to `ts` files). 

Thanks to VSCode, adding types to typescript files based on JSDOC (=transformed files) is automated and extremely simple so we have that to our advantage which is **GREAT** (see the PR description for more info).
Meaning that the combination of the transformer and VSCode is very powerful.

It's going to save lots of time and energy and allow us to focus on the essence of the migration.

People who want to help with the migration can really focus on types and stuff instead of syntax and replacing comments.
This will allow us to migrate and keep existing PRs (at least most of them I hope) which was a big concern of ours.

Take a look at a fantastic example of types by @asturur [here](https://github.com/fabricjs/fabric.js/pull/8019#discussion_r903555173)

**Comment by bhabgs** [comment]
I support using typescript and hope to support the Chinese market better. Because I'm from China....

**Comment by ShaMan123** [comment]
### Hello community,
TS is coming 🤩

## A Quick Update
The old build system has been replaced with `rollup` #8013 
Now we focus on migrating to TS. 
The team, @asturur @melchiar and I @ShaMan123, decided for! It is a done fact. 
The first milestone would be moving to es6 not caring about types and focusing on import/export.
After that it's a type festival.
The es6 step is a bit tricky so the team is doing that (contributions are still welcome, contact/ping us if you want in).
The second step is the main event for TS fans, the type-fest 🥳. 
Everyone is welcome aboard. We will decide how to manage that later on.

## Why I am writing this update
We are reaching out to the community after a POC.
We have tested migrating to classes (which is granted) but **we encountered 2 hurdles that must be mitigated**.
Or else `class` will be disregarded as unwanted 😥.

**We have considered some approaches and would like to hear your ideas and considerations.**

## Guidelines for the solution
1. code must be simple and easy
1. **NO** duplicate values/variables
1. **NO** ugly assignments
1. support a way to subclass/override these values with ease at runtime (this part is a bit flexible in terms of when and how the overriding is done)
2. untranspiled/transpiled code **MUST** behave similarly
3. TS compatible - type inference should work out of the box

## (1) Default values
fabric uses the prototype of a shape (`Object`, `Rect` etc.) to remove default values when exporting to json, see [here](https://github.com/fabricjs/fabric.js/blob/5bf76d4a72020df15e8605e02d732df0c4e6a949/src/shapes/object.class.js#L893).
The problem is that TS class properties are not set on the prototype. They are assigned to the instance in the constructor during initialization, build #8101 to see for your self.

![image](https://user-images.githubusercontent.com/34343793/182214277-50119e0d-9ffa-4b24-894d-0cbd6158c36d.png)

So there isn't a way to get default values from the class.

### Thoughts
- assign defaults to prototype in class static init (failing 2, 3, 4)
- export a static method `getDefaultValues` (failing 2, 3 if we are honest)
- deprecate `removeDefaultValues` and change `toObject(propsToInclude)` to populate **ONLY** `propsToInclude`, ignoring the rest (I prefer this approach)

### Tryouts
- generating files with default values during build #8105 

## (2) Subclassing a superclass
Let's say you are a dev using fabric@v9 a year from now. Everything is there, typescript, classes (maybe even webgl, who knows?!).
You need to override a method defined on `fabric.Object` but you want the change to propagate to all classes.

**How do we support that?**

### Thoughts
We could convert all subclasses to [mixins](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#mix-ins)
```ts
export function RectMixinGenerator<T extends new (...args: unknown[]) => unknown>(Klass: T) {
    return class RectMixin extends Klass {
        //   rect class body
    }
}

```
And then you could create custom objects by using the mixin with your subclass of `Object`

```ts

class ObjectOverride extends fabric.Object {
    someMethod() {
       // override logic
    }
}

export const Rect = RectMixinGenerator(ObjectOverride);
```

But it seems TS is finding it hard to digest this approach #8100 (see `applyMixins`)

**Comment by avra-m3** [comment]
Hey @ShaMan123 ,
Happy to lend a hand on smaller tasks outside of work hours, I'll send you an email.

The typescript migration for a library like fabric is never going to be perfect, I think class mixins will be fine but will likely leave a lot of existing applications on fabric 5, there should be a plan to support those applications with security updates, etc in the medium term if that's the path we go down. 
Regarding the first point, given the migration is already going to be a nightmare I believe you should aim for the best solution rather than the one that makes the migration the simplest. In our use case we've never used `removeDefaultValues` and leaned heavily on `toObject(propsToInclude)`.

Look forward to seeing the results!

**Comment by Lazauya** [comment]
@ShaMan123
I'm having trouble understanding the requirements about overriding during runtime, specifically the requirement "support a way to subclass/override these values with ease at runtime (this part is a bit flexible in terms of when and how the overriding is done)" 

Does this mean that for (1) you should be able to set the default values during run time? Like I should be able to set the default border radius of a rectangle, and then that becomes true for all rectangles?

For (2), does this mean that you wish to be able to override a method like `getTotalAngle()` and have all rectangles use your new method?

(1) might be achievable using TS decorators inspired by [this Stackoverflow post]( https://stackoverflow.com/a/32440846/3210986). [This](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAZwKZQAoCc5TgCgDcBDAGxFQC5FiwBPASkQG8AoRDxLdELMffFGJYA5umq06AGkQBrVHWrIoWGGBFMAvAD4W7TgaGj0AbXl0Auok2IS5VAG59HAL4MnL1qwiliyZIgAQn6oehwQCMpYINBwWPhMbK6snt6+-ogAGoioAB5QqGAAJgHBaGGIAAJomDh4+ABEABIAogAybQDyDQz6uQD8SipqIk76EWBRMXjxic4oIAAOqLNOyak+fgEAmjn5hSVZFdXo2LgEAMwATL0cdIOIYCAAtgBGK2PhkSrTcQkVHGQSxWCTWiE8G2+iDo1keqAA7ohtqDvJE4KRUAA6UhwET4OiY3IyAmMBxAA) is a more a complete example. My understanding is that if you want to override stuff at runtime, you'd just need to modify the prototype.

**Comment by ShaMan123** [comment]
@Lazauya yes to both of your questions.
The [decorators](https://www.typescriptlang.org/docs/handbook/decorators.html) approach is interesting (but is still experimental, we must take that into consideration). In total is viable and good and meets all guidelines.

>  My understanding is that if you want to override stuff at runtime, you'd just need to modify the prototype.

[TS Playground Link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAZwKZQAoCc5TgCgDcBDAGxFQC5FiwBPASkQG8AoRDxLdELMffFGJYA5umq06AGkQBrVHWrIoWGGBFMAvAD4W7TgaGj0AbXl0Auok2IS5VAG59HAL4MnL1qwiliyZIgAQn6oehwQCMpYINBwWPhMbK5eHABGCdRgIAC2qahYiAA+iOAAJqjAaqilYZzcULxIAKxOyZ7evv6IABqIqAAeUKhgpQHBaLUAAmiYOHj4AEQAEgCiADJrAPILDPr9APxKKmoiTvoRYFExePGJziggAA75Ca2Inu0+fgEAmn2Dw1GPSmM2wuAIAGYAEy7Dh0Q6ILK5fJncKRFTXOIJWocZBPF7ufTtNIZRE5PIFYplCpVGpJOo8PiIFpE1jtH4AOkec1wdGeHNS1kQoEgsAQggAFjBkJQfncOPVGogoFLkBy6B4vONUFyeVA+TrBTYRdB4PwVdLZfKuIykAA2TXedGIOhCsCoADuiB+rydlzgpB1pDgInwdA5-Rk4ekLoFCXcQA) shows you are right regarding modifying the prototype.

**Comment by ShaMan123** [comment]
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier

I wonder what happened after v3.7

**Comment by Lazauya** [comment]
@ShaMan123 
> modifying a prototype doesn't pass down to subclasses due to the strict nature of class.

Actually, I think this behavior is favorable. I didn't even know this until just now, but it actually DOES override in the subclasses if you haven't overridden that function explicitly in the class definition, which I think makes sense. See [this example](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAZwKZQAoCc5TgCgDcBDAGxFQC5FiwBPASkQG8AoRDxLdELMffFGJYA5umq06AGkQBrVHWrIoWGGBFMAvAD4W7TgaGj0AbXl0Auok2IS5VAG59HAL4MnL1qwiliyZIgAQn6oehwQCMpYINBwWPhMbK5eHABGCdRgIAC2qahYiAA+iOAAJqjAaqilYZzcULxIAKxOyZ7evv6IABqIqAAeUKhgpQHBaLUAAmiYOHj4AEQAEgCiADJrAPILDPr9APxKKmoiTvoRYFExePGJziggAA75Ca2Inu0+fgEAmn2Dw1GPSmM2wuAIAGYAEy7Dh0Q6ILK5fJncKRFTXOIJWocZBPF7ufTtDgAehJiAAIpsAHIAFUQcEI+VU5UQUAAFqFQJBYAhEJzuPoyYh0gxMjk8gVimUKlUaklEMKOPVGogWkLyR8vF8ugAtf5DEa-WoXK6xW44h7PW5vYmK8kgCYc0J460AOlFMgA7qEILQUKhQjAoAByAKoJlIRnMmDlJAC0JezlIH2IbJwUowYB0RCPOa4OjPfSi8XIqUlEaysDVS2mqAikJgUuS6xW-IehIAQjeyp4fDViAA1A20GBbax2sKfm68+CoIXUB7W9zoPB+ByYMhKD87kquH2kBvkG66E5hZ90YgczZq17ED9Xt5InBSIvSHARPg6G7+jJv9Irw7BhCVYYVNiZLAWROEVEDUIIQhoEZEH6WCkF6YhoBAMhSBzCB2VoMQAmDVC2U5B5Uh1NBkFA8lAhAesaTWP4s1I0J52eRAvT8REI3yBkIJZOMRQqOJUFYcZF1nPB2MXVJl3AVcEEEdlN23O5ewafsADYPFYboZ3zGSf1bBYAHFNk2ClAh+FZOwWJw6yvKFW1ve9H0cgAvFzUDvXV3OfV83XfT9v1-QCAO-T1EA8oDCSAA).

I mean, I guess I just don't know a use case where you would have a base class like (lets call it `Base`) and a derived class (called `X`) where you have a different definition for some function `f` in `Base` and `X`. It doesn't make sense to me that changing the prototype of `Base` would change `X.f` too if `X` has a completely different definition of `f`. It make more sense that you would simply change the the prototype of `X` directly if you wanted to make them both the same.

**Comment by ShaMan123** [comment]
@Lazauya I agree with all you wrote.
Excellent, thanks.
@asturur take a look

**Comment by Lazauya** [comment]
So what's the word here?

**Comment by asturur** [comment]
The only thing i want to investigate so i get a grasp of it is why it behavea differently.
Mdn says, class definition is just a syntactic sugar on top of functions. FabricJs probably with createClass was doing some extra magic that i m happy to remove as soon as i understand it fully.
If functions do not support prototype override at runtime i l fine with that

**Comment by Meesayen** [comment]
Hi, I was thinking of a possible solution to (1) Default Values, and thought we could maybe quickly create a dummy instance of the same class as the object, during the `_removeDefaultValues` function execution and then compare the two, e.g.:

```js
   // from this
    _removeDefaultValues: function(object) {
      var prototype = fabric.util.getKlass(object.type).prototype;
      ...
    }
    
    // to this
    _removeDefaultValues: function(object) {
      var prototype = Reflect.construct(object.constructor, [])
      ...
    }
```

Now the only problem I see is that some classes may have mandatory constructor arguments, and in that case that `Reflect.construct` line would throw unless we pass the necessary arguments list :/ For Fabric classes might be an easy problem to fix, AFAIK most of the classes either have an optional `options` object, or an `element` + and optional `options`, but for custom classes in the wild, maybe we'll need to ask users to provide a `getDefaultConstructorParams` which I guess breaks guideline 3.

**Comment by ShaMan123** [comment]
At this point I think we should drop `removeDefaultValues`. The dev can easily override classes once #8203 in done.
As mentioned in previous comments, proto assignments are possible (without inheritance).
It will be breaking and migrating will be an effort.
But it will be worth it.

**Comment by jiayihu** [comment]
From the consumer POV, one of the biggest headaches with using fabric in our company has the heavy reliance on custom behaviours in the class implementation. Just as an example, I've found sometimes that using `callSuper("originalMethod")` in an overridden method results in unexpected results compared with `originalMethod.call(this)` because of how `callSuper` actually works.

So as @Lazauya , I'm in favour of following the strict standard, even if it can result in more boilerplate code or extra work, as far as it keeps the behaviour consistent with the standard.

Regarding the transition to TS, it's late for a feedback, but it would be very welcome from us. I've actually often seen minor errors or inconsistencies in the fabric codebase (when overriding) that would have been avoided if TS were used internally as well. Actually, sometimes errors in the manual `d.ts` files have required us extra work to workaround it, such as aliasing a method to a different name so that we can give it the correct type.

**Comment by Lazauya** [comment]
@asturur 
> The only thing i want to investigate so i get a grasp of it is why it behavea differently.

I made a TS Playground here to better understand what extra stuff fabric was doing [here](https://www.typescriptlang.org/play?#code/PTACBcGcFoDsHsDGALApog1gKAG4EMAnAAkgBsBLRVIgXiIEECC8BPAOgAcD5weWPUbMpVQAaIqgC2HcCwBiAV1iJw5eLFpEAZkpVrYACgCURAN5EAvqKxFbRYACoHRUOQDmCAtQfAbdvAAmAQCyqODI8AGQmjrKquoGGKR4kJDikPAKBFTiHISosOAmpn522vDEBvjEXPACBLJE5BoZWVTFpWVEiOoZpIKk8G4GAIziSSlpJJnZYkR5XoW53PWyRgDcnWU9sH0DQ6NsY0RbXWdltassTRoTqZzcvLIC4qfn58+o8FpEd5APPD4AgA2pdUA0WABdV7vWF2AytWaglbg2SQogAaiIAHJsUYNm8mj8DGCITdfsl7rUnvxqAAyOmEsqfb4UyYAmkg0lo2g0OjY2J6dTYogMpnwxFUZF1VFQzE4vFsZoBVAADwA8loDNjEHhSKQAMoKep4ogAPiI0BGHThOz2bEGwwATOJwuQpn90jMcvN8kt5iiIQS4eKQGyqY8gahpVd0XQDIL4oZuSwbbDQ8A7Hb4P0HQcAMyu5Du8aUqaSuYLArgZYyoObDN2LzgLIaRP6YxmRvbXo5-bDAAsRZLRDd-zt4AIChUFWH496k+nvAIQmN4MQZdLky9bUrfprAbrsh3SJT0KIhDcCkk1cgwdhYbK1RIa+yZc0Y7YE6nM5XkFfG6TJscKPnYn7fkuFSrvUgGpJoVaFMBD6Zk+hBEM2rYAGp6go1B0BWMaypCbB4BwHCkCwBhjuIl7Xre97vKBtjgQuP7LtB67vvhAFlg2IEoXY5DEimRAAIR8jizTkKoerkAAXqgpolPx5wYQQGhqbA2GkLhSGMQJtgWN2lh8chlhGCSgZrKZ7x-ByUaERCcbaLoSadspcLdL2uaOgYQ6jsWUwsbsi6-nOX6sZBf48du4blt6e6LAeKYnlKZ40QQV43oUd56bCz7-jBXEBe6EUhWxUGFZxQHisxgVlZAoXsVVb6TPB+55e8z6adpuGaARZ4kWRFFUYFGVZfRnUfPVEG-hxrVwdxRU1bVRJEJZR7XOJ-JSTJFAKUpq22Jp6FhFhOGoFNZxGXCFhXZYhKoKQkDUB5tllvZnyOTy+EJd9UL3TdXRA4ZDbtuoRBGgARrBkDuQ9WDgxour6ka9QGDe4SRAAcngN5ps+CHgKEWMBJosAKPq1hnAA+mOH6BUh2Y+Qc1G+klJMRAE95hhu5CSFDh6ctQKB4M0o7wNoypEJjXNEAg4ASAAjgoeoS90er9Ip0Sy5EfgAO7Fv0610zNkVzS1sNprYBWvgAwmWnORJopulbNzUxRGgJfbrAS4zekJ5UJJtjsCvv+6g6LbS+9QO5MTvc12ZxEwn-X247YRc1NUNeHgGB5SD9iZqgwiFNAATungUP9HAargNAFCwKgnSu9EdCtw1TWVZ7-zUlGSFGX4wcGKJKeZ5E1unS26lebsfZ5sM2KTuQqBk7wGuo6+OLyuHePUFi2LiL78s8OUShk+LfefN0yBi7Ah8lblfiD8dZ0zwYtHZVADoFG44TmkQa0nQAD87NqwJyGuRSibNhBUC-JrD+mU6I5XENaIwnQABcYDCgQJRqQUa7oCSD0cA4GwzgAAST16jlGIIgXOSYiCsgAESwxLkwtgZCXA3gFuCTU2gq4EEoGwBQqhSCcNAAsPGZhFBxH0BYIgoJ9zojqEmNWTC46pCYerZoaBBGKy0NwSQ4jJGSDMOqKGAArdA4B5HfRXpAdEAAFKy9iSC3y8GTKG1xNY3EangZQJdGE-HprDThnkzgGChiLfIOhSAy0iEJFgzQ3CMMsdY6IKotDNFXkQXR1AUiPyIPrTIpAyZ4C0FoaxF59S+PAP4qgd4yG+CRt0ehqANFwwJmhIm5NKakGpl0bkrj8IUDgXgxBE0UFEAAAwEiHsSFkPwhkl2BNMuMEkBSuX0IdC4+54IuJLkIYsWhwDGAHn4Fpfx3KdE-LtcgskFKQJGmzT+k1n5YD8HZS2xUiZIS+QoGGZZAl0GBIHD5thh5E0ntDWGn1aTtSSnCgQeU7JX3hXQJu+tIYAthmczoRNVyAsmIcjgChIDIESGWe8QMtAVHWs+cgmhpniH6LAP+yB9lHnsT-Nl4R1hNCIAAHiIKy9l-LyAYgxJPZm-YDAAHYnTynIOIcxViVBsAwKgFgcMibAKRdQYBoDTAWCMFuVIZre6Rk+LWVY9jgTkHPFCvKgQQjjyiJS2KyzID2sdfual8z1qiVRVa2kSpYDSXuftVAk9g3e1DXch5eEJDSFkDIoUsBzm2FjcLTuFViB0D+H8j6aKBDwM3tQugeC0bgiQidQtWBB4TmxfUAAQikJNdDUB4HAO0ssBgPKoA4JQBOWCWnwxlQvAwTC0AsCtEwjYD0TUfKbW2l6mhO3dt7ZMAw1aCCrrmAOodiAR0uVkRDeG4FNa7qnYO4dbr538onb5adWqH2LrmU2gAGuutpHSDD7oGeYW9x63Wjq2QkYoj8y2GlfDeo9Cc31PoOC+lgIw2AAFZ52WEsDCMwRBVRgbPRocd3lZUoadG++RVg-DmAI6e9N61INIeGCh-MlGBm2BYIRhjJG54sxYzOgc7HOhyW4ww3j9pn1VtfJRnD6CsBNtVOTVAWLP14tVGwYDCc8XMYMBprTbr0EadVOptglECQabkmcoAA)

In short, `createClass` really isn't doing much. It's literally just a more verbose and less safe version of the normal ES6 classes.
It handles mixins by progressively merging all the objects passed into it. 
The only thing notable is how it handles it handles super methods. If you use `callSuper` somewhere in a method you define, it will wrap the method in another method that sets the `superclass` property so that `callSuper` can work recursively in super methods, and won't get confused. It checks for `callSuper` by looking at the text of the function, and I feel like you can confuse it somehow; ES6 `super` doesn't have that problem.

As far as I can tell ES6 classes should literally just be drop-in-able if you replace all the calls to `callSuper`. It needs a new mixin system too but I think @ShaMan123 did that already.

**Comment by ptgamr** [comment]
@Lazauya @ShaMan123 

> (1) you should be able to set the default values during run time? Like I should be able to set the default border radius of a rectangle, and then that becomes true for all rectangles?

Just want to chime in here, in our application this behaviour creating issues as we have two different canvases behave quite differently in the same view - setting the default values in one canvas will have unexpected consequences in the other canvas and vice versa. 

I and think I don't mind creating my own classes to provide my own default values, ie.

```ts
class MyRectWithSomeDefaults extends fabrics.Rect {
  constructor() {
    // set the default values in here
  }
}
```

**Comment by atmikev** [comment]
I'm relatively new to fabricJS and am glad to see its getting an update. I'm a little confused on why you would want default values for base objects though. @ptgamr is right that subclassing would be the right way to do that.

**Comment by asturur** [comment]
https://github.com/fabricjs/fabric.js/pull/8246

Expanded my concerns a bit here in this PR with some example code ( this is a discussion pr is not being merged ro anything like that ).

Yes subclassing is a good way to do it, you can do it, this is not stopping subclassing at all.
But also:
- subclassing is not for everyone and a lot of people got to the same end result chaning the values on the prototype
- subclassing Text and having it stay in the middle of the prototype chain between Object and IText or Textbox is even harder
- You don't have to do prototype changes if you don't feel like, it is just an alternative less verbose approach

**Comment by asturur** [comment]
If i want to add an extra property to the cacheProperties, or remove one, if i want to change originX/originY for my project to center/center, or if i want to work with objects with default stroke to 0, and if i want to assign a different control set to a class, i don't want to force a developer into subclassing. It was never necessary, it shouldn't be necessary now.

**Comment by asturur** [comment]
> @Lazauya @ShaMan123
> 
> > (1) you should be able to set the default values during run time? Like I should be able to set the default border radius of a rectangle, and then that becomes true for all rectangles?
> 
> Just want to chime in here, in our application this behaviour creating issues as we have two different canvases behave quite differently in the same view - setting the default values in one canvas will have unexpected consequences in the other canvas and vice versa.
> 
> And I think I don't mind creating my own classes to provide my own default values, ie.
> 
> ```ts
> class MyRectWithSomeDefaults extends fabrics.Rect {
>   constructor() {
>     // set the default values in here
>   }
> }
> ```

In particular which default values where you setting that were bleeding over the other canvas?

**Comment by asturur** [comment]
> The only thing notable is how it handles it handles super methods. If you use `callSuper` somewhere in a method you define, it will wrap the method in another method that sets the `superclass` property so that `callSuper` can work recursively in super methods, and won't get confused. It checks for `callSuper` by looking at the text of the function, and I feel like you can confuse it somehow; ES6 `super` doesn't have that problem.
> 
> As far as I can tell ES6 classes should literally just be drop-in-able if you replace all the calls to `callSuper`. It needs a new mixin system too but I think @ShaMan123 did that already.

Yes es6 classes are drop-in repleaceable, and the mixins can be added as before, assigning them over the prototype, no changes are necessary, and the resulting class is perfectly valid

**Comment by ShaMan123** [comment]
I prefer [chaining mixins](https://github.com/fabricjs/fabric.js/blob/49afc8e4c88fac0a319a14d973d0704c0f908569/src/mixins/apply_mixins.ts) with a mixin builder (e.g. [Collection](https://github.com/fabricjs/fabric.js/blob/49afc8e4c88fac0a319a14d973d0704c0f908569/src/mixins/collection.mixin.ts)) as mentioned in [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#mix-ins).
That way you can subclass Text and recreate IText with your subclass.
Sure it requires a few more lines of code.
But it releases fabric from a patched and confusing js standard. 
That is the whole point of the migration. So to me it seems fit that we become strict in complying to it and drop what doesn't fit anymore.

**Comment by asturur** [comment]
chainging mixins is doable, it changes nothing, but what you see as confusing and patched ( patched why? the JS defines the prototype inheritance in a certain way, and classes leverages it entirely ) may not be for other developers.
Extending a class vertically and/or horizontally is just a matter of what fit best with the model.

Rect is a superset of Object, while for example SVGexport is a functionality you can add to any of the classes or not, without changing what the class is. a Rect stays a Rect, without the need to become a `SVGEnanchedClass extends BaseClass`.


Is also possible that this can be done at lower level doing:

```js
Rect.__proto__ = SvgExtender(Rect.__proto__) 
```

so adding it programmatically to already existing classes, but for the scope of how fabricJS code is written vs how is used, nothings change.

The fact that when i open the VSCode editor i want to see the code organized in a certain way that conform with my liking is something i had to put away when i became the `not only maintainer anymore`.

So i try to stick to measurable changes, and everyone should do the same.
- how much code
- which functionalities
- easy of use
- enabling new patterns vs breaking the old one
- eventually speed and correctness

**Comment by ShaMan123** [comment]
Could you define the next steps as we see it?
I would prefer seeing a PR you've done to understand the new standard.
Or make a list of dos/donts

**Comment by asturur** [comment]
i can convert object for the sake of seeing how it would look like.
The next step would be pull out what we lose if we don't add public fields in the fabricJS codebase and evaluate.
From my small deep dive, we don't lose anything.
Which are the different opinions?

**Comment by ShaMan123** [comment]
I suggest leaving Object to the end if possible. It will be complex and messy with all the mixins and logic broken all across fabric, some even shared with canvas.
I would choose Rect or Circle, the simplest possible, wondering if it will be a viable POC if Object remains as is (??).

I am not fully grasping your point, forgive me. I will need to see code and review and ask and discuss then.
Sounds to me that some of these fields (e.g. `cacheProperties`) should become static.
Defaults of course can't.

**Comment by ShaMan123** [comment]
sounds good
naming should be changed for sure
Shape?

**Comment by Lazauya** [comment]
I vote `FabricObject`. I think it's the most clear and will be the least confusing when used in other codebases.

**Comment by asturur** [comment]
Yesterday i prepared the code but i got stuck in one of those problem where our mixed import strategy makes now object undefined and i have no idea how to fixed. Working on it.

**Comment by asturur** [comment]
https://github.com/fabricjs/fabric.js/pull/8322

Ok this is a PR.
Go with the negatives please, pointing out issues ( that are not cited already in the pr description )

**Comment by ideaviewes** [comment]
please provide a version of react native

**Comment by ShaMan123** [comment]
@ideaviewes it is out of scope for this thread.
I suggest you open a dedicated discussion.

**Comment by fallenartist** [comment]
I'm sorry for bumping into this adult conversation but I thought I might represent a voice of us little folk that: a) are enjoying Fabric and are hoping to utilise it more in the future, and b) are not coding ninjas and sometimes struggle with more advanced concepts of JS. I do hope that moving in whatever direction will not mean making using Fabric any less problematic for non-expert programmers like me and there will be enough examples and well written docs that it will be possible to use it without a need to hire a pro or rush to SO with every little issue. I've had this experience with D3 where it became somewhat harder and harder to use recently (but that might well be the little old brain of mine!). Thanks and good luck!

**Comment by SLKnutson** [comment]
Just a minor thing, you can use `ts-expect-error` instead of `ts-ignore`. The difference is that `ts-expect-error` won't compile if the next line is not a typescript error. That way, when the underlying issue is fixed, you will automatically notice that the issue is fixed and typing can be re-enabled for that line.

**Comment by ShaMan123** [comment]
Hi everyone,

A year has passed since this ticket was opened.

**[Beta is out!](https://github.com/fabricjs/fabric.js/releases/tag/v6.0.0-beta1)** 
Published last week, just in case anyone missed it.
You can install it via 
```bash
npm i fabric@beta
```

We migrated the entire codebase and made significant changes and fixes, see #8299 and the [CHANGELOG](https://github.com/fabricjs/fabric.js/blob/master/CHANGELOG.md).
Most of the codebase is fully typed. Some complex modules remain in need of typing and perhaps more touch ups. 

Building is done with rollup.
fabric has its own cli now so you can enter `npm run cli -- -h` and continue from there.
You may experience stuff with consuming types and/or the build, submit an issue if you do. We are still figuring that out.

Repository docs have been updated as well to reflect v6 changes (README, CONTIBUTION etc.). More work is needed in this field. The website remains outdated. 
This will be the next major effort after we release v6. We will announce updates about that in the future.

Remember it is beta, things aren't final and may change (the entry point especially).
Considering the scope of changes, regressions might surface.
Also take into consideration a shift in the way fabric is used. Subclassing is now super easy (class is no. 1 citizen in v6) and there is no fabric namespace.

As always, contributions are welcome (either TS, website or anything else). Ping one of us in an issue/discussion/PR so we can direct the efforts and give a hand or suggest a solution.

**Comment by bladerunner2020** [comment]
With T or without, but we are badly missing some types in v6. For example, BaseFabricObject, TEvent, ObjectEvents, etc.

**Comment by ShaMan123** [comment]
@bladerunner2020 tracker issue #8868

**Comment by urschrei** [comment]
Is there a new way to access the `.fabric` property of a `Canvas` element in the beta?

This

```javascript

import {Canvas, Image} from 'fabric';

const canvas = new Canvas('certcanvas');
document.getElementById('certcanvas').fabric = canvas;
Image.fromURL('assets/img/cert.png', function(img) {
    img.set({
        lockMovementX: true,
        lockMovementY: true,
        selectable: false
    });
    img.scaleToHeight(566);
    img.scaleToWidth(800);
    canvas.add(img);
```

Doesn't seem to work for me…

**Comment by ShaMan123** [comment]
fabric is now typed - use that.
I wouldn't wait for the website, unless you join the effort and give a hand.

**Comment by ShaMan123** [comment]
For the greater community:

**We want to wrap up the TS migration and we need help.**

**Contributions are needed!**

Current task is to type the remaining files and address TS errors.
Contributors can search for files with `@ts-nocheck` and get to work.
Open a draft PR with the title `chore(TS): [SCOPE]` once you get started stating which files you wish to address so people don't work on the same thing. 
Then when you are done, move it out of the draft state and review it and then one of us will review as well.
Be sure to check for open PRs before you start working.
If someone wishes to do it all go ahead.

Thanks!

_Originally posted by @ShaMan123 in https://github.com/fabricjs/fabric.js/issues/8299#issuecomment-1644418966_

**Comment by asturur** [comment]
Ok we are ready to close this.
The last `ts-nocheck` have been removed, now the library is fully typed.

We are doing a last check up of the api we changed during this year of migrations and we are releasing a stable 6.0.
We will handle missing exported types as patches ( 6.0.x ) but we will try to catch all the missing exports before.

Eraser brush is not ready yet and is going to be added as the next thing.
We are also working on docs and we will try to update all the introductions to fabricJS on the website with modern content that make sense
Some of them have been written in 2011 and are outadated.
